// supercolliderjs/lib/dryads/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.layer = exports.SynthEventList = exports.SynthStream = exports.SynthControl = exports.SCSynthDef = exports.AudioBus = exports.Synth = exports.Group = exports.SCLang = exports.SCServer = undefined;
exports.dryadic = dryadic;
exports.play = play;
exports.h = h;

var _scserver = require('./middleware/scserver');

var _scserver2 = _interopRequireDefault(_scserver);

var _SCServer = require('./SCServer');

var _SCServer2 = _interopRequireDefault(_SCServer);

var _SCLang = require('./SCLang');

var _SCLang2 = _interopRequireDefault(_SCLang);

var _Group = require('./Group');

var _Group2 = _interopRequireDefault(_Group);

var _Synth = require('./Synth');

var _Synth2 = _interopRequireDefault(_Synth);

var _AudioBus = require('./AudioBus');

var _AudioBus2 = _interopRequireDefault(_AudioBus);

var _SCSynthDef = require('./SCSynthDef');

var _SCSynthDef2 = _interopRequireDefault(_SCSynthDef);

var _SynthControl = require('./SynthControl');

var _SynthControl2 = _interopRequireDefault(_SynthControl);

var _SynthStream = require('./SynthStream');

var _SynthStream2 = _interopRequireDefault(_SynthStream);

var _SynthEventList = require('./SynthEventList');

var _SynthEventList2 = _interopRequireDefault(_SynthEventList);

var _dryadic = require('dryadic');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// re-export all the Dryad classes
exports.SCServer = _SCServer2.default;
exports.SCLang = _SCLang2.default;
exports.Group = _Group2.default;
exports.Synth = _Synth2.default;
exports.AudioBus = _AudioBus2.default;
exports.SCSynthDef = _SCSynthDef2.default;
exports.SynthControl = _SynthControl2.default;
exports.SynthStream = _SynthStream2.default;
exports.SynthEventList = _SynthEventList2.default;

// export the layer for app = dryadic().use(layer)


// confusing to swap the names like this
/**
 * @module dryads
 * 
 */

var middleware = [_scserver2.default];

var classes = [_SCServer2.default, _SCLang2.default, _Group2.default, _Synth2.default, _AudioBus2.default, _SCSynthDef2.default, _SynthControl2.default, _SynthStream2.default, _SynthEventList2.default];

var layer = exports.layer = {
  middleware: middleware,
  classes: classes
};

/**
 * Create a DryadPlayer from a Dryad or hyperscript definition.
 *
 * Automatically includes the supercollider.js layer
 *
 * usage:
 *
 *   var sc = require('supercolliderjs');
 *   var player = sc.dryadic([
 *     'scserver', [
 *       ['group', [
 *         ['synth', {
 *           defName: 'sinosc',
 *           args: {
 *             freq: 440
 *           }
 *         }]
 *       ]
 *   ]);
 *   player.play();
 *   ...
 *   player.stop();
 */
function dryadic(rootDryad) {
  var moreLayers = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : [];
  var rootContext = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};

  return (0, _dryadic.dryadic)(rootDryad, [layer].concat(moreLayers), rootContext);
}

/**
 * Play a Dryad or hyperscript document.
 *
 * usage:
 *
 *   var sc = require('supercolliderjs');
 *   var player = sc.play([
 *     'scserver', [
 *       ['group', [
 *         ['synth', {
 *           defName: 'sinosc',
 *           args: {
 *             freq: 440
 *           }
 *         }]
 *       ]
 *   ]);
 *
 * @param {Dryad|Array} rootDryad - Dryad object or hyperscript document
 * @returns {DryadPlayer}
 */
function play(rootDryad) {
  return dryadic(rootDryad).play();
}

/**
 * Convert hyperscript object to a tree of Dryads.
 *
 * This lookups each class by lower class 'classname'
 * and creates an instance with properties and children.
 */
function h(hgraph) {
  var player = dryadic();
  return player.h(hgraph);
}
