// supercolliderjs/lib/dryads/SCLang.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dryadic = require('dryadic');

var _sclang = require('../lang/sclang');

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var defaultOptions = {
  debug: true,
  echo: false,
  stdin: false
};

/**
 * Boots a new SuperCollider language interpreter (sclang) making it available for all children as context.sclang
 *
 * Always boots a new one, ignoring any possibly already existing one in the parent context.
 *
 * `options` are the command line options supplied to sclang (note: not all options are passed through yet)
 * see {@link lang/SCLang}
 *
 * Not to be confused with the other class named SCLang which does all the hard work.
 * This Dryad class is just a simple wrapper around that.
 */

var SCLang = function (_Dryad) {
  _inherits(SCLang, _Dryad);

  function SCLang() {
    _classCallCheck(this, SCLang);

    return _possibleConstructorReturn(this, (SCLang.__proto__ || Object.getPrototypeOf(SCLang)).apply(this, arguments));
  }

  _createClass(SCLang, [{
    key: 'defaultProperties',
    value: function defaultProperties() {
      return {
        options: defaultOptions
      };
    }
  }, {
    key: 'prepareForAdd',
    value: function prepareForAdd() {
      return {
        callOrder: 'SELF_THEN_CHILDREN',
        updateContext: function updateContext(context, properties) {
          return {
            sclang: (0, _sclang.boot)(_lodash2.default.defaults(properties.options, { log: context.log }))
          };
        }
      };
    }
  }, {
    key: 'remove',
    value: function remove() {
      return {
        run: function run(context) {
          return context.sclang.quit();
        }
      };
    }
  }]);

  return SCLang;
}(_dryadic.Dryad);

exports.default = SCLang;
