// supercolliderjs/lib/dryads/SynthEventList.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _dryadic = require('dryadic');

var _Group = require('./Group');

var _Group2 = _interopRequireDefault(_Group);

var _msg = require('../server/osc/msg');

var _iterators = require('./utils/iterators');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * Takes a list of synth event objects with relative times and schedules them.
 *
 * ## properties
 *
 * __events:__ Array
 *
 * The event values should be simple JavaScript objects:
 *
 *     {
 *       defName: 'synthDefName',
 *       args: {
 *         out: 0,
 *         freq: 440
 *       },
 *       time: 0.3
 *     }
 *
 *  Where time is seconds relative to the epoch. The epoch is the start time of
 *  the dryadic tree, unless a parent Dryad has set a new epoch into context.
 *
 *    epoch: number|Date|undefined
 *      Optional epoch that the event times in the list are relative to.
 *      Can also be updated by the updateStream
 *      default: context.epoch or now
 *
 * __updateStream:__ Bacon stream to push updated event lists of the form:
 *
 *      {
 *        events: [{time: msgs: []}...],
 *        epoch: 123456789
 *      }

 *     .events Array
 *     .epoch  number|Date
 *
 * Deprecated: will be replaced with live updating and setting of
 * Any value in a dryadic document from the player or remote client.
 *
 * Pushing a new event list cancels previous events and schedules new events.
 *
 * Note that by default the epoch will be unchanged: relative times
 * are still relative to when the Dryad tree started playing or when any parent
 * Dryad set an epoch in context. This means you update the currently playing score
 * but it doesn't restart from the beginning, it keeps playing.
 *
 * Optionally you may push an .epoch with the updateStream. This can be a date or timestamp
 * slightly in the future. If you pass "now" then any events at `0.0` will be too late to play.
 *
 * __defaultParams:__ a fixed object into which the event value is merged.
 *
 * __loopTime:__ Play the events continuously in a loop.
 */
var SynthEventList = function (_Dryad) {
  _inherits(SynthEventList, _Dryad);

  function SynthEventList() {
    _classCallCheck(this, SynthEventList);

    return _possibleConstructorReturn(this, (SynthEventList.__proto__ || Object.getPrototypeOf(SynthEventList)).apply(this, arguments));
  }

  _createClass(SynthEventList, [{
    key: 'add',

    /**
     * @param  {DryadPlayer} player
     * @return {Object}      Command object
     */
    value: function add(player) {
      var _this2 = this;

      var commands = {
        scserver: {
          schedLoop: function schedLoop(context, properties) {
            // temporary: we need to know the play time of the whole document
            var epoch = context.epoch || _lodash2.default.now() + 200;
            if (epoch !== context.epoch) {
              context = player.updateContext(context, { epoch: epoch });
            }

            return _this2._makeSchedLoop(properties.events || [], properties.loopTime, epoch, context);
          }
        }
      };

      // built-in stream support will be added to Dryadic
      // for now it is hard to detect Bacon.Bus as being an object,
      if (this.properties.updateStream) {
        commands = _lodash2.default.assign(commands, {
          run: function run(context, properties) {
            var subscription = properties.updateStream.subscribe(function (streamEvent) {
              var ee = streamEvent.value();
              var loopTime = _lodash2.default.isUndefined(ee.loopTime) ? properties.loopTime : ee.loopTime;
              var epoch = ee.epoch || context.epoch || _lodash2.default.now() + 200;
              if (epoch !== context.epoch) {
                context = player.updateContext(context, {
                  epoch: epoch
                });
              }

              player.callCommand(context.id, {
                scserver: {
                  // need to set epoch as well because OSCSched uses that for relative times
                  schedLoop: function schedLoop(ctx /*, props*/) {
                    return _this2._makeSchedLoop(ee.events || [], loopTime, ctx);
                  }
                }
              });
            });

            player.updateContext(context, { subscription: subscription });
          }
        });
      }

      return commands;
    }
  }, {
    key: '_makeSchedLoop',
    value: function _makeSchedLoop(events, loopTime, context) {
      var synthEvents = this._makeMsgs(events, context);
      return loopTime ? (0, _iterators.loopedEventListIterator)(synthEvents, loopTime) : (0, _iterators.eventListIterator)(synthEvents);
    }
  }, {
    key: '_makeMsgs',
    value: function _makeMsgs(events, context) {
      var defaultParams = this.properties.defaultParams || {};
      return events.map(function (event) {
        // TODO: do this a jit time in the schedLoop
        var defName = event.defName || defaultParams.defName;
        var args = _lodash2.default.assign({ out: context.out || 0 }, defaultParams.args, event.args);
        var msg = (0, _msg.synthNew)(defName, -1, _msg.AddActions.TAIL, context.group, args);
        return {
          time: event.time,
          msgs: [msg]
        };
      });
    }

    /**
     * @return {Object}  command object
     */

  }, {
    key: 'remove',
    value: function remove() {
      var _this3 = this;

      return {
        run: function run(context) {
          if (context.subscription) {
            if (_lodash2.default.isFunction(context.subscription)) {
              // baconjs style
              context.subscription();
            } else {
              // Rx style
              context.subscription.dispose();
            }
          }
        },
        scserver: {
          sched: function sched(context) {
            // unschedAll
            return _this3._makeSchedLoop([], context.epoch, context);
          }
        }
      };
    }

    /**
     * @return {Dryad}  Wraps itself in a Group so all child Synth events will be removed on removal of the Group.
     */

  }, {
    key: 'subgraph',
    value: function subgraph() {
      return new _Group2.default({}, [this]);
    }
  }]);

  return SynthEventList;
}(_dryadic.Dryad);

exports.default = SynthEventList;
