// supercolliderjs/lib/index.js
'use strict';

var _server = require('./server');

var server = _interopRequireWildcard(_server);

var _lang = require('./lang');

var lang = _interopRequireWildcard(_lang);

var _map = require('./map');

var map = _interopRequireWildcard(_map);

var _msg = require('./server/osc/msg');

var msg = _interopRequireWildcard(_msg);

var _resolveOptions = require('./utils/resolveOptions');

var _resolveOptions2 = _interopRequireDefault(_resolveOptions);

var _checkInstall = require('./utils/checkInstall');

var _checkInstall2 = _interopRequireDefault(_checkInstall);

var _Errors = require('./Errors');

var _scapi = require('./scapi');

var _scapi2 = _interopRequireDefault(_scapi);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

/**
 * @module supercolliderjs
 */
module.exports = {
  server: server,
  lang: lang,
  map: map,
  msg: msg,
  resolveOptions: _resolveOptions2.default,
  checkInstall: _checkInstall2.default,
  SCError: _Errors.SCError,
  SCLangError: _Errors.SCLangError,
  scapi: _scapi2.default,
  // @deprecated These were renamed
  sclang: lang,
  scsynth: server
};
